### PR TITLE
Sub Rogue - Base Rotation improvements

### PR DESCRIPTION
**Pull request by TsaboTavok**
Created some of the most important Sub Rogue analysis for the base rotation. 
Refactored a lot of the existing Sub code to use Ability Tracker from core.

### CODE REVIEW

File: src/Parser/Rogue/Subtlety/Modules/BaseRotation/NightbladeDuringSymbols.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    const filtered = this.symbolsDamageTracker.getAbility(SPELLS.NIGHTBLADE.id);
+
+    const badRefreshShare = filtered.casts / total.casts;
+    when(badRefreshShare).isGreaterThan(0)
+    .addSuggestion((suggest, actual, recommended) => {
+      return suggest(<Wrapper>Do not refresh <SpellLink id={SPELLS.NIGHTBLADE.id} /> during <SpellLink id={SPELLS.SYMBOLS_OF_DEATH.id} /> - cast <SpellLink id={SPELLS.EVISCERATE.id} /> instead. You can refresh <SpellLink id={SPELLS.NIGHTBLADE.id} /> early to make sure that its up for the full duration of <SpellLink id={SPELLS.SYMBOLS_OF_DEATH.id} />. </Wrapper>)
+        .icon(SPELLS.NIGHTBLADE.icon)
+        .actual(`You refreshed Nightblade ${filtered.casts} times during Symbols of Death.`)
+        .recommended(`0 is recommend.`)
+        .regular(0.1).major(0.2);
```
Comment (Pewtro) [inline src/Parser/Rogue/Subtlety/Modules/BaseRotation/NightbladeDuringSymbols.js]: same as above, the recommended should be updated.

File: src/Parser/Rogue/Subtlety/Modules/CastTracker/FilteredDamageTracker.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  }
+
+  on_byPlayer_cast(event) {
+    if(!this.shouldProcessEvent(event)) return;
+    super.on_byPlayer_cast(event);
+  }
+    
+  shouldProcessEvent(event) {
+    return false;
+  }
```
Comment (Pewtro) [inline src/Parser/Rogue/Subtlety/Modules/CastTracker/FilteredDamageTracker.js]: Instead of having this redirect unnecessarily, why don't you just move the onByPlayer_damage into the two classes that extend this one, and just have those extend the DamageTracker? You never call this one anyways - seems overly redundant to me.

File: src/Parser/Rogue/Subtlety/Modules/RogueCore/SymbolsDamageTracker.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+class SymbolsDamageTracker extends FilteredDamageTracker {
+  static dependencies = {
+    combatants: Combatants,
+  };
+  
+  shouldProcessEvent(event) {
+    return this.combatants.selected.hasBuff(SPELLS.SYMBOLS_OF_DEATH.id);
+  }
+}
+
```
Comment (Pewtro) [inline src/Parser/Rogue/Subtlety/Modules/RogueCore/SymbolsDamageTracker.js]: The two damage tracker are fine as is, but you could probably merge them and have two shouldProcessEvent calls called something like shouldProcessEventSymbols and shouldProcessEventDance or whatever, not a needed change, but maybe something to consider.

File: src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowContribution.js [hunk 4]
```diff
@@ -1,31 +1,24 @@
...
+    const danceDamage = Object.keys(this.danceDamageTracker.abilities)
+    .map(abilityId => this.danceDamageTracker.abilities[abilityId].damageEffective || 0)
+    .reduce((a,b) => a+b,0) * 0.3 / 1.3;
+     
     return (
         <StatisticBox
           icon={<SpellIcon id={SPELLS.DARK_SHADOW_TALENT.id} />}
           value={`${formatNumber(danceDamage / this.owner.fightDuration * 1000)} DPS`}
           label="Increased from Dark Shadow talent"
-          tooltip={`Total damage increase is ${formatNumber(danceDamage)} in ${this.shadowDance.totalShadowDanceCast} Shadow Dance casts.`}
```
Comment (Pewtro) [inline src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowContribution.js]: the label should probably read "damage contribution from Dark Shadow", makes more sense to me.

File: src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowEvis.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  statistic() {    
+  const totalEviscerateHitsInShadowDance  = this.danceDamageTracker.getAbility(SPELLS.EVISCERATE.id).damageHits;
+    const danceEvis = totalEviscerateHitsInShadowDance / this.totalShadowDanceCast;
+    return (
+        <StatisticBox
+          icon={<SpellIcon id={SPELLS.EVISCERATE.id} />}
+          value={`${danceEvis.toFixed(2)}`}
+          label="Average Eviscerates in Shadow Dance"
+          tooltip={`Your average Eviscerate casts per Shadow Dance. Your cast ${totalEviscerateHitsInShadowDance} Eviscerates in ${this.totalShadowDanceCast} Shadow Dances. This number includes Eviscerates cast from Death from Above. Subtlety rogue should cast as many as possible (usually 2 times) Eviscerates in a Shadow Dance to get benefit from 30% damage increasing of Dark Shadow talent.`}
+          />
```
Comment (Pewtro) [inline src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowEvis.js]: If you have a inclusion of "Subtlety rogue should cast as many as possible (usually 2 times) Eviscerates in a Shadow Dance to get benefit from 30% damage increasing of Dark Shadow talent." in the tooltip, then it seems clear that this should instead be a suggestion for this specific module.

File: src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowNightblade.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+class DarkShadowNightblade extends DarkShadow {
+  suggestions(when) {
+    const nightblade = this.danceDamageTracker.getAbility(SPELLS.NIGHTBLADE.id).casts;
+    when(nightblade).isGreaterThan(0)
+    .addSuggestion((suggest, actual, recommended) => {
+      return suggest(<Wrapper>Do not cast <SpellLink id={SPELLS.NIGHTBLADE.id} /> during <SpellLink id={SPELLS.SHADOW_DANCE.id} /> when you are using <SpellLink id={SPELLS.DARK_SHADOW_TALENT.id} />. </Wrapper>)
+        .icon(SPELLS.NIGHTBLADE.icon)
+        .actual(`You cast Nightblade ${nightblade} times during Shadow Dance.`)
+        .recommended(`0 is recommend.`)
+        .major(0.5);
```
Comment (Pewtro) [inline src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowNightblade.js]: update the recommended as per the other ones further above.

File: src/Parser/Rogue/Subtlety/Modules/BaseRotation/CastsInShadowDance.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    + this.danceDamageTracker.getAbility(SPELLS.EVISCERATE.id).casts;
+
+    const missedCastsInDanceShare = castsInDance/(danceCount * 4);
+    when(missedCastsInDanceShare).isLessThan(1)
+    .addSuggestion((suggest, actual, recommended) => {
+      return suggest(<Wrapper>Try to cast 4 spells during each <SpellLink id={SPELLS.SHADOW_DANCE_BUFF.id} /> </Wrapper>)
+        .icon(SPELLS.SHADOW_DANCE_BUFF.icon)
+        .actual(`You cast less ${castsInDance} spells during Shadow Dance out of ${danceCount * 4} possible.`)
+        .recommended(`${danceCount * 4} recommended`)
+        .regular(0.90).major(0.80);
```
[the remark this answers is not in the document]
Comment (TsaboTavok) [reply src/Parser/Rogue/Subtlety/Modules/BaseRotation/CastsInShadowDance.js]: The idea here is that you must cast exactly 4 times (max possible), but I don't want to have it major if its just one dance with 3. So I am checking the average casts to find out how bad it is but suggesting exactly 4.

File: src/Parser/Rogue/Subtlety/Modules/CastTracker/FilteredDamageTracker.js (hunk 2 above)
Comment (TsaboTavok) [reply src/Parser/Rogue/Subtlety/Modules/CastTracker/FilteredDamageTracker.js]: because it  would lead to duplicating the logic for filtering 3 types of events to both existing trackers. I don't understand what you mean by "You never call this one anyways" - its overridden in both FilteredDamageTracker and DanceDamageTracker

File: src/Parser/Rogue/Subtlety/Modules/BaseRotation/CastsInShadowDance.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Pewtro) [reply src/Parser/Rogue/Subtlety/Modules/BaseRotation/CastsInShadowDance.js]: It would read better to say 4 casts per shadow dance is recommended, or something similiar. Giving me a multiplied number as recommended, doesn't really help me on the fly - whereas "4 pr cast" is quite tangible. 
The wording in "actual" is still bad and needs to be updated regardless.

File: src/Parser/Rogue/Subtlety/Modules/RogueCore/SymbolsDamageTracker.js (hunk 3 above)
Comment (TsaboTavok) [reply src/Parser/Rogue/Subtlety/Modules/RogueCore/SymbolsDamageTracker.js]: I defiantly want them separate, because I'm using the base DamageTracker functionality to aggregate all data. Having 1 tracker with multiple aggregates would be similar in usage, but harder to maintain in my opinion.

File: src/Parser/Rogue/Subtlety/Modules/CastTracker/FilteredDamageTracker.js (hunk 2 above)
Comment (Pewtro) [reply src/Parser/Rogue/Subtlety/Modules/CastTracker/FilteredDamageTracker.js]: Do you care about the heal events of those spells? Do they even have heal events, and of what use are they?

Comment (TsaboTavok) [reply src/Parser/Rogue/Subtlety/Modules/CastTracker/FilteredDamageTracker.js]: I don't care about heal events, but the idea here is to filter out all events. Would make no sense to have a tracker that calculates everything during a certain condition, but ignores that condition for healing. I can disable healing aggregation by creating 3  "shouldProcess" functions and not overriding the healing one. Even if I drop healing its still casts and damage filtering that must be duplicated to every tracker. And I will need more - I will need more of these trackers down the line. My other argument is that I find SymbolsDamageTracker to be more concise when it only has 1  method overridden - shouldProcessEvent, instead of having duplicated logic between all trackers.

File: src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowEvis.js (hunk 5 above)
Comment (TsaboTavok) [reply src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowEvis.js]: This was an old statistic in this module before I picked it up. I'm still on the fence about it, its defiantly not a suggestion - the proper suggestions have been created in this PR (for example cast 4 spells during dance). However it can be a useful statistic just like "Dark Shadow damage contribution"  to see why you are performing worse then someone else. I was planing to get some extra opinions on it and then decide to remove or not.

Comment (Pewtro) [reply src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowEvis.js]: It should 100% be a suggestion and not in the tooltip.

Comment (TsaboTavok) [reply src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowEvis.js]: But its not a valid suggestion. Its sometimes a mistake to cast 2 Eviscerates in  Dance. I can just remove it now if it helps

Comment (TsaboTavok) [reply src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowEvis.js]: Just not sure why it has to be handled right now considering its old and unrelated to this PR

File: src/Parser/Rogue/Subtlety/Modules/BaseRotation/CastsInShadowDance.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  }
+  suggestGloomblade(when) {
+    const filtered = this.danceDamageTracker.getAbility(SPELLS.GLOOMBLADE_TALENT.id);
+    when(filtered).isGreaterThan(0)
+    .addSuggestion((suggest, actual, recommended) => {
+    return suggest(<Wrapper>Use <SpellLink id={SPELLS.SHADOWSTRIKE.id} /> instead of <SpellLink id={SPELLS.GLOOMBLADE_TALENT.id} /> during <SpellLink id={SPELLS.SHADOW_DANCE.id} />. </Wrapper>)
+        .icon(SPELLS.GLOOMBLADE_TALENT.icon)
+        .actual(`You cast Gloomblade ${filtered} times during Shadow Dance.`)
+        .recommended(`0 is recommend.`)
+        .regular(0).major(1);
```
[the remark this answers is not in the document]
Comment (Pewtro) [reply src/Parser/Rogue/Subtlety/Modules/BaseRotation/CastsInShadowDance.js]: Still not saying ``is recommended`` which is the grammatical correct.

File: src/Parser/Rogue/Subtlety/Modules/BaseRotation/CastsInShadowDance.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Pewtro) [reply src/Parser/Rogue/Subtlety/Modules/BaseRotation/CastsInShadowDance.js]: the wording is still off. Now you're suggestion 4 times shadowDance casts per shadow Dance.

File: src/Parser/Rogue/Subtlety/Modules/BaseRotation/NightbladeDuringSymbols.js (hunk 1 above)
Comment (Pewtro) [reply src/Parser/Rogue/Subtlety/Modules/BaseRotation/NightbladeDuringSymbols.js]: Not saying ``is recommended`` as needed.

File: src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowNightblade.js (hunk 6 above)
Comment (Pewtro) [reply src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowNightblade.js]: not saying ``is recommended``

File: src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowEvis.js (hunk 5 above)
Comment (Pewtro) [reply src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowEvis.js]: Because you're adjusting it in a PR, you might as well fix it properly. If it's majorly handled by something newly added, just remove it or reword it to correctly showcase what you want it to showcase. 
This could be something like ``You should cast as many Eviscerates as necessary to get four total spell casts in Shadow Dance``.

Comment (TsaboTavok) [reply src/Parser/Rogue/Subtlety/Modules/Talents/DarkShadow/DarkShadowEvis.js]: Ok I've added suggestion because I want to keep it in for now so I can get some feedback on it.

**Comment by TsaboTavok** [comment]
Do you need any additional changes/info?

**Comment by Pewtro** [comment]
Looks good to me, should be everything I could find in this one - otherwise you'll have a bug or two to fix in the near future. 
Please think about the suggestion from sref/kfinch going forward in terms of dealing with the trackers and maybe put them in, in a future PR.

**Comment by TsaboTavok** [comment]
Thanks, I will continue to refactor the code as I work on it, so changing thinks to fabricated events is something I will defiantly try.
